redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+///
+/// The `ScanState` struct represents the state of a scan operation in a Redis cluster.
+/// It holds information about the current scan state, including the cursor position, scanned slots map,
+/// address being scanned, and address's epoch.
+
+const BITS_PER_U64: usize = u64::BITS as usize;
+const NUM_OF_SLOTS: usize = SLOT_SIZE as usize;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub(crate) struct ScanState {
+    // the real cursor in the scan operation
+    cursor: u64,
+    // a map of the slots that have been scanned
+    scanned_slots_map: SlotsBitsArray,
+    // the address that is being scanned currently, based on the next slot set to 0 in the scanned_slots_map, and the address "own" the slot
+    // in the SlotMap
+    pub(crate) address_in_scan: String,
+    // epoch represent the version of the address, when a failover happens or slots migrate in the epoch will be updated to +1
+    address_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    pub(crate) scan_state_cursor: ScanStateCursor,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+}
+
+#[derive(Debug, Clone)]
+/// Represents the type of an object in Redis.
+pub enum ObjectType {
+    /// Represents a string object in Redis.
+    STRING,
+    /// Represents a list object in Redis.
+    LIST,
+    /// Represents a set object in Redis.
+    SET,
+    /// Represents a sorted set object in Redis.
+    ZSET,
+    /// Represents a hash object in Redis.
+    HASH,
+}
+
+impl std::fmt::Display for ObjectType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            ObjectType::STRING => write!(f, "string"),
+            ObjectType::LIST => write!(f, "list"),
+            ObjectType::SET => write!(f, "set"),
+            ObjectType::ZSET => write!(f, "zset"),
+            ObjectType::HASH => write!(f, "hash"),
+        }
+    }
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_cursor: ScanStateCursor,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<ObjectType>,
+    ) -> Self {
+        Self {
+            scan_state_cursor,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateCursor {
+    scan_state_ref: Arc<Option<ScanState>>,
+}
+
+impl ScanStateCursor {
+    /// Creates a new instance of `ScanStateCursor` from a given `ScanState`.
+    fn from_scan_state(scan_state: &ScanState) -> Self {
+        Self {
+            scan_state_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateCursor`.
```

Comment:
I think for the end user its is more simple if starting a new scan is `ScanStateCursor::new` than `ScanStateCursor::default`
Might be better to impl both

Reply:
you're missing the usage of default.
If all of a struct's fields implement default, then the struct can derive default without any effort. So implementing default always enables your users more than `new` does.

Reply:
But default is part of the macros, why usn't it enogh?

Reply:
what macros? default is a trait.

Reply:
`#[derive(Debug, Clone, Default)]
/// A wrapper struct for managing the state of a scan operation in a cluster.
/// It holds a reference to the scan state and provides methods for accessing the state.
/// The `status` field indicates the status of the scan operation.
pub struct ScanStateRC {
    scan_state_rc: Arc<Option<ScanState>>,
    status: ScanStateStage,
}`

Since iv'e derived Default in my macros, the default for this struct will be -
`ScanStateRC {
    scan_state_rc: Arc<None>,
    status: Initiating,
}`
Since default of option is none and i decided that Initiating is the default of the status.
Using new and default will give you the same result. But new is more clear to the user.